flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/stream/StreamExecRank.java

```diff
@@ -0,0 +1,217 @@
+import org.apache.flink.configuration.ConfigOptions;
+import org.apache.flink.streaming.api.operators.KeyedProcessOperator;
+import org.apache.flink.streaming.api.transformations.OneInputTransformation;
+import org.apache.flink.table.api.TableConfig;
+import org.apache.flink.table.api.TableException;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.planner.codegen.EqualiserCodeGenerator;
+import org.apache.flink.table.planner.codegen.sort.ComparatorCodeGenerator;
+import org.apache.flink.table.planner.delegation.PlannerBase;
+import org.apache.flink.table.planner.plan.nodes.exec.ExecEdge;
+import org.apache.flink.table.planner.plan.nodes.exec.ExecNode;
+import org.apache.flink.table.planner.plan.nodes.exec.ExecNodeBase;
+import org.apache.flink.table.planner.plan.utils.AppendFastStrategy$;
+import org.apache.flink.table.planner.plan.utils.KeySelectorUtil;
+import org.apache.flink.table.planner.plan.utils.RankProcessStrategy;
+import org.apache.flink.table.planner.plan.utils.RetractStrategy$;
+import org.apache.flink.table.planner.plan.utils.UpdateFastStrategy;
+import org.apache.flink.table.runtime.generated.GeneratedRecordComparator;
+import org.apache.flink.table.runtime.generated.GeneratedRecordEqualiser;
+import org.apache.flink.table.runtime.keyselector.RowDataKeySelector;
+import org.apache.flink.table.runtime.operators.rank.AbstractTopNFunction;
+import org.apache.flink.table.runtime.operators.rank.AppendOnlyTopNFunction;
+import org.apache.flink.table.runtime.operators.rank.ComparableRecordComparator;
+import org.apache.flink.table.runtime.operators.rank.RankRange;
+import org.apache.flink.table.runtime.operators.rank.RankType;
+import org.apache.flink.table.runtime.operators.rank.RetractableTopNFunction;
+import org.apache.flink.table.runtime.operators.rank.UpdatableTopNFunction;
+import org.apache.flink.table.runtime.typeutils.InternalTypeInfo;
+import org.apache.flink.table.types.logical.LogicalType;
+import org.apache.flink.table.types.logical.RowType;
+
+import java.util.Collections;
+import java.util.stream.IntStream;
+
+/**
+ * Stream {@link ExecNode} for Rank.
+ */
+public class StreamExecRank extends ExecNodeBase<RowData> implements StreamExecNode<RowData> {
+
+	// It is a experimental config, will may be removed later.
+	@Experimental
+	public static final ConfigOption<Long> TABLE_EXEC_TOPN_CACHE_SIZE =
+		ConfigOptions.key("table.exec.topn.cache-size")
+			.longType()
+			.defaultValue(10000L)
+			.withDescription("TopN operator has a cache which caches partial state contents to reduce" +
+				" state access. Cache size is the number of records in each TopN task.");
+
+	private RankType rankType;
+	private RankProcessStrategy rankStrategy;
+	private RankRange rankRange;
+	private boolean generateUpdateBefore;
+	private boolean outputRankNumber;
+	private int[] partitionFields;
+	private int[] sortFields;
+	private boolean[] sortDirections;
+	private boolean[] nullsIsLast;
```

Comment:
mark them as final. and I would like to suggest reorder the fields as: type part, partition part, order by part, range part, and other part. which could make read easier to understand, because it matches the order in sql definition.